### PR TITLE
CiviEvent - Error registering participants via search task

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
This aims to fix smarty fatal error on using a search-task to register contacts for an event. Steps to reproduce:

* Perform an advanced search
* Select one or more contacts
* Choose the register contacts to an event search task

Before
----------------------------------------
Form does not display. There's a Smarty error.

After
----------------------------------------
Form displays. No error.

Comment
----------------------------------------
This is an expansion on the prior draft #19118. It is working for me but I think warrants a better fix in master (separate the forms)

### CODE REVIEW

File: CRM/Event/Form/FormBlock.php [hunk 1]
```diff
@@ -0,0 +1 @@
+<?php
```
Comment (seamuslee001) [inline CRM/Event/Form/FormBlock.php:1]: @eileenmcnaughton why is this needed?

Comment (eileenmcnaughton) [reply CRM/Event/Form/FormBlock.php:1]: arg - the extra files would be part of the bigger fix - but the 2 small changes you might like to pull into your PR & I'll close this

**Comment by totten** [comment]
OK, when I `r-run` this, it's definitely better. :) It creates the registration, and there's no hard crash.

But...

I fill out the form like so:

Screen Shot 2020-12-04 at 10 07 16 PM

The first time, it submitted but displayed this error:

Screen Shot 2020-12-04 at 10 04 55 PM

Which was weird. I figure maybe a fluke or some other hackery on my local, so I tried again. Again, it creates the registration (yay) but produces an error:

Screen Shot 2020-12-04 at 10 07 26 PM

The second error seems to be more stable - i.e. it showed up again on the 3rd and 4th tries.

**Comment by totten** [comment]
There's more discussion/debugging [in Mattermost log](https://chat.civicrm.org/civicrm/pl/pafnzncyqtf8uxta4rukd5iq4o).

@eileenmcnaughton I've pushed up a change which (1) ensures the `action` is ADD (*my system was donig VIEW*), (2) calculates the `qfKey` for the ADD subform (`CRM_Event_Form_Participant`), (3) reinstates `qfKey` enforcement, and (4) removes the extra `qfKey`.

However, I don't think we can get further review+backport tonight. (We'll punt backport to 5.32.2.)

File: CRM/Event/Page/Tab.php [hunk 2]
```diff
@@ -90,7 +90,7 @@ public function edit() {
     $controller = new CRM_Core_Controller_Simple(
       'CRM_Event_Form_Participant',
       'Create Participation',
-      $this->_action, FALSE, FALSE, TRUE
```
Comment (eileenmcnaughton) [inline CRM/Event/Page/Tab.php]: The removal of this TRUE has caused the 500 console error that I added it to avoid to come back. The other symptom is the fees not loading.

Comment (totten) [reply CRM/Event/Page/Tab.php]: That's weird. I was seeing the opposite - even with `ignoreKey=TRUE` here (PR-branch without df80fba), the AJAX call replied 500 and the fees subform didn't render. With df80fba (with a valid key -- and with `ignoreKey=FALSE`), the AJAX call and the subform worked.

I wonder if there might be some subtle difference in the pageflows or use-cases that we're trying?

Comment (eileenmcnaughton) [reply CRM/Event/Page/Tab.php]: @totten maybe - I have a good idea of a better fix - but we need to get 'something' out - to be honest I can't understand how the key would be rejected if ignoreKey is false

I am doing a basic search & then selecting some contacts & choosing register participants

I did test that on the pr test site & it failed

Comment (totten) [reply CRM/Event/Page/Tab.php]: FWIW, my procedure is very similar, ie

* Open advanced search
* Search for all contacts (no criteria)
* Check the first "Individual" in the demo data ("Adams, Heidi")
* In the task list, choose to register
* For the event, choose any of the three demo events (eg "Rain-forest Cup")

@eileenmcnaughton I can reproduce the error that appears on the test site. (Incidentally, I'm doing this on a different workstaiton/build.) Notably, in both test site and local, the 500 error is not related to qfKey. It's about "getFieldValue failed" (which curiously [arose before](https://github.com/civicrm/civicrm-core/pull/19125#issuecomment-739133215) but had seemed transitory). Here's a backtrace.

Screen Shot 2020-12-07 at 9 22 41 PM

The thing is... to my reading, it's basically saying that [the event-registration form's preProcess()](https://github.com/civicrm/civicrm-core/blob/df80fbabc873519be8b08d2d49a682826e671994/CRM/Event/Page/Tab.php#L121-L130) is applying a mismatched security check. (It's trying to use a `contactId` and `CRM_Contact_Page_View::checkUserPermission()` -- but in the context of a search-task, it doesn't make sense to use a singular `contactId`  for assessing permissions on a bulk action.)

I've pushed up another patch which fixes this symptom for me, but it definitely needs some more eyes to see if it's faithful to contract/permissioning of this route/controller.

Comment (eileenmcnaughton) [reply CRM/Event/Page/Tab.php]: @totten given that 
1) the person who has reported this has applied the patch in the 5.32 branch & says it is working and
2) we are pretty clear none of this is the final fix

I'm inclined to merge what that person merged & work on the right fix in master

Comment (totten) [reply CRM/Event/Page/Tab.php]: The report on 5.32 PR actually pointed to the same problem raised here (just a diff part of the manifestation -- the HTTP 500 error you noted above is the "getFieldValue required" error) - I think they just didn't realize that the screen was misbehaving (because it's a misbehavior of omission; because the immediate error-message is hidden in AJAX console; and the UI-level error message is delayed). So the revision they tested isn't actually working.

Added some more [comments over there](https://github.com/civicrm/civicrm-core/pull/19127#issuecomment-741170965) to try to distinguish 3 issues floating around here. To my testing, the current revision of this branch addresses all 3.

**Comment by totten** [comment]
Did a final `r-run` on autobuild test site. Tested both the search-task (registering 1x and 3x contactx) and the standalone menu item ("Events => Register event participants" with 1x contact). In all cases, the full form displays. No AJAX HTTP 500. No lingering error-notices. 

Merging this as near-term fix per MM discussion. Eileen's aiming to do a cleaner/separate route for this in `master`.
